### ISSUE TITLE
PPA for easy updating

### ISSUE THREAD

**Issue by eligrey** [antelle/keeweb#141]
I'd like to self-host, but I'm probably going to forget to update this occasionally due to the effort involved. A launchpad PPA would make it as easy as running `apt-get update && apt-get upgrade` whenever I want to update KeeWeb.

**Comment by antelle** [comment]
That would be great but I need some help with creating a binary package and updating it.
The same for arch linux: #78

**Comment by vincent-t** [comment]
@antelle I have created a script wich builds a .deb package.
Save as ```buildkeeweb-desktopdeb.sh```:
```
#!/bin/bash
echo "Build keeweb-desktop.deb"
rm -rf keeweb-desktop
rm -f keeweb-desktop.deb
mkdir -p keeweb-desktop/opt/keeweb-desktop
wget -nc https://github.com/antelle/keeweb/releases/download/v1.0.4/KeeWeb.linux.x64.zip
unzip KeeWeb.linux.x64.zip -d keeweb-desktop/opt/keeweb-desktop
mkdir -p keeweb-desktop/usr/bin
tee "keeweb-desktop/usr/bin/KeeWeb" << 'EOF'
#!/bin/bash
/opt/keeweb-desktop/KeeWeb --disable-updater $1
EOF
chmod 755 keeweb-desktop/usr/bin/KeeWeb
mkdir keeweb-desktop/DEBIAN
tee "keeweb-desktop/DEBIAN/control" << 'EOF'
Package: keeweb-desktop
Version: 1.0.4
Section: base
Priority: optional
Architecture: amd64
Maintainer: antelle <antelle.net@gmail.com>
Installed-Size: 97280
Homepage: https://keeweb.info/
Description: KeeWeb
 Free cross-platform password manager
 compatible with KeePass
EOF
mkdir -p keeweb-desktop/usr/share/applications
tee "keeweb-desktop/usr/share/applications/keeweb.desktop" << 'EOF'
[Desktop Entry]
Name=KeeWeb
GenericName=Password Manager
GenericName[de]=Passwortverwaltung
GenericName[es]=Gestor de contraseñas
GenericName[fr]=Gestionnaire de mot de passe
GenericName[ru]=менеджер паролей
Comment=Free cross-platform password manager compatible with KeePass
Exec=KeeWeb %u
Icon=keeweb
Terminal=false
StartupNotify=false
Type=Application
Categories=Utility;
MimeType=application/x-keepass2;
StartupWMClass=KeeWeb
EOF
mkdir -p keeweb-desktop/usr/share/mime/packages
tee "keeweb-desktop/usr/share/mime/packages/keeweb.xml" << 'EOF'
<?xml version="1.0" encoding="UTF-8"?>
<mime-info xmlns="http://www.freedesktop.org/standards/shared-mime-info">
  <mime-type type="application/x-keepass2">
     <comment>KeePass 2 database</comment>
     <glob pattern="*.kdbx"/>
  </mime-type>
</mime-info>
EOF
mkdir -p keeweb-desktop/usr/share/icons/hicolor/128x128/apps
cp keeweb-desktop/opt/keeweb-desktop/resources/app/icon.png keeweb-desktop/usr/share/icons/hicolor/128x128/apps/keeweb.png
fakeroot dpkg-deb --build keeweb-desktop
echo "keeweb-desktop.deb successfully build!"

```

**Comment by antelle** [comment]
Thanks! I've made a grunt tasks from this and uncluded building deb in grunt build, now it's producing deb file as well. Registered the app: https://launchpad.net/~antelle-net/+archive/ubuntu/keeweb 
I'm building the package on os x using dpkg from brew and I get this error when trying to install it on ubuntu in virtualbox:
![ubuntu 15 running 2016-04-02 11-24-43](https://cloud.githubusercontent.com/assets/633557/14225283/c08096e6-f8c5-11e5-9378-3b1e3b3a896e.png)
What is the reason, any guess?

**Comment by vincent-t** [comment]
@antelle As you can see you need fakeroot. Without it the ownership is wrong (not root). Try to install fakeroot via brew (brew install fakeroot).

**Comment by antelle** [comment]
Yes, I've tried fakeroot from brew and standalone installation: both does not work on el capitan :(

**Comment by vincent-t** [comment]
Maybe this helps: https://github.com/Homebrew/homebrew/issues/33400. Or you could try to use ```sudo chown -R root:root keeweb-desktop``` before ```dpkg-deb --build keeweb-desktop```. (I have not tested it on my system).

**Comment by antelle** [comment]
running sudo from build is not the best idea:
1. files permissions will be changed and clean will require sudo as well
2. to fix 1, running entire build with sudo is even worse
3. it will not work in some IDEs, e.g. webstorm

I've found some packages in npm which create deb righ from node.js, without using dpkg - will try using them.

**Comment by antelle** [comment]
So, I switched the build to grunt-contrib-deb and after some patches (you may find them in my fork here: https://github.com/antelle/grunt-contrib-deb) it's installing. Could you please help me with these questions, I still don't understand it:
1. After installation, there's no way to launch the app: no desktop icons or search results in Ubuntu search. I guess some files are missing here. Which ones?
2. I've installed the package, is it ok, that there's no uninstall button in Ubuntu Software Center, or I should add something to the package for it to appear?
3. It shows 'License: Unknown' and 'Updates: Unknown' - will it be updated automatically with apt-get?

**Comment by vincent-t** [comment]
This is very strange with my build script the produced deb (build on ubuntu 14.04) installs without these problems. I have an desktop icon and the option to uninstall the package. If I'm back home I'll take a look at these issues.

**Comment by antelle** [comment]
Ok, thank you. Attached the contents of PRM (with extracted content.tar.gz data.tar.gz). Maybe there's something wrong with permissions?..
[contents.txt](https://github.com/antelle/keeweb/files/200804/contents.txt)

**Comment by antelle** [comment]
So, I've investigated a bit:
1. Search in apps: it was permissions trouble; after adding another patch to packager, it appears in apps search. I also added [a postinstall script](https://github.com/antelle/keeweb/blob/develop/package/deb/scripts/postinst) to refresh icons cache. Is it a right way of doing this?
2, 3. Still don't understand

**Comment by vincent-t** [comment]
Ok, 2 is very strange because with my script it is working out of the box.
In the next few days I'll try to investigate this further. 3. Updates only
work via ppa or manual Installation of a newer version, so this is expected
behavior.
Am 02.04.2016 9:43 nachm. schrieb "antelle" <notifications@github.com>:

**Comment by antelle** [comment]
So the package appears in Ubuntu Software Center for you after installation?
I'm using Ubuntu 15. I can upload the deb with it...

**Comment by vincent-t** [comment]
I checked your package and it is not working because the rights are wrong (see attached files).
You can also remove sudo in your postinstall script.

[yourdeb.txt](https://github.com/antelle/keeweb/files/206319/yourdeb.txt)
[mydeb.txt](https://github.com/antelle/keeweb/files/206321/mydeb.txt)

**Comment by antelle** [comment]
Thanks, I'll add rights

**Comment by antelle** [comment]
I've changed rights, here's the new package: https://www.dropbox.com/s/pqyfjy055kjsl03/keeweb-desktop_1.1.1-1_amd64.deb?dl=0
Does it work for you?

**Comment by vincent-t** [comment]
@antelle It is now working. :) Only bug left the description is not shown because the formatting is wrong. Here is the right formatting:
```
Description:
 Free cross-platform password manager
 compatible with KeePass
```

**Comment by antelle** [comment]
What's the correct way to build it from string? Just wrap all lines to fit N characters per line rule? I cannot find any info about such splitting [in debian policy manual](https://www.debian.org/doc/debian-policy/ch-binary.html#s-descriptions).
You mean, the first line here is a synopsis and the last one is the extended description?

**Comment by vincent-t** [comment]
It is not the splitting. First line is the package name, second line needs a space + description text. If you put the description in the fist line nothing is displayed. ;)

```
Description: Name
 Description
```

**Comment by antelle** [comment]
So is it ok to format is in this way?
```
Description:
 Free cross-platform password manager compatible with KeePass
```

**Comment by vincent-t** [comment]
Yes and in the first line you could add the application name

```
Description: KeeWeb
 Free cross-platform password manager compatible with KeePass
```

**Comment by antelle** [comment]
Implemented this, the package is installing for my ubuntu 15 as well.
What's the next step, is there any manual how to push this package? I haven't found any good description. [Uploading](https://help.launchpad.net/Packaging/PPA/Uploading) page describes only how to upload the source to ppa (if I understood what they mean correctly), but how can I upload the binary deb?

**Comment by vincent-t** [comment]
Sorry I have no experience with ppa's. As far as I understand you need to upload the source code or the source for the deb file. The ppa builds the packages from source (https://askubuntu.com/questions/71199/manually-created-deb-how-do-i-upload-to-a-ppa).

**Comment by antelle** [comment]
Bad, so there will be no option to upload this, as we definitely don't want to rebuild the app from source. I'll try to search whether it possible to add the app to Ubuntu Software Center.

**Comment by tagplus5** [comment]
https://github.com/tagplus5/keeweb-ppa PPA for ubuntu based on your .deb packages.

**Comment by peteruithoven** [comment]
So is that ppa officially supported? Are there still things that need to happen?

**Comment by antelle** [comment]
> So is that ppa officially supported?

I don't support it, but as I can understand, it's regularly updated.

**Comment by likewei92** [comment]
I'm on Solus right now and installing manually. Seeing that this is already on launchpad, is there a possibility of publishing it as a snap, to cater to a wider range of distros?

https://snapcraft.io/docs/build-snaps/ci-integration

**Comment by antelle** [comment]
It's possible of course, but will require some preparation: #450
